File: py_common/mpfr_base32.py
"""
MPFR Base-32 Conversion Utilities

Functions for converting between gmpy2 mpfr objects and MPFR base-32 string format.
Uses gmpy2 library which provides Python bindings to MPFR, ensuring exact compatibility
with the C implementation.
"""

import gmpy2
from gmpy2 import mpfr  # type: ignore


def remove_trailing_zeros(s: str) -> str:
    """
    Remove trailing zeros from a number string that has a decimal point.
    Also removes the decimal point if no fractional part remains.
    
    Args:
        s: Number string
    
    Returns:
        String with trailing zeros removed
    """
    if '.' not in s:
        return s
    
    # Remove trailing zeros
    s = s.rstrip('0')
    # Remove decimal point if it's now at the end
    s = s.rstrip('.')
    return s


def parse_mpfr_base32(s: str, precision: int = 256) -> mpfr:
    """
    Parse an MPFR base-32 string to a gmpy2 mpfr object.
    Supports formats:
    - Plain: "1a" or "-1a"  
    - With decimal point: "1.a" or "-1.a" or "0.00001a"
    
    Args:
        s: Base-32 string to parse
        precision: Precision in bits for the mpfr object (default: 256)
    
    Returns:
        gmpy2 mpfr object
    """
    # Set precision context
    with gmpy2.context(precision=precision):  # type: ignore
        # gmpy2 can directly parse base-32 strings using mpfr_set_str
        # The format expected by MPFR is the same as our input
        try:
            result = mpfr(s, base=32)
            return result
        except:
            # Fallback: handle special cases or errors
            if s == '0':
                return mpfr(0)
            raise ValueError(f"Cannot parse base-32 string: {s}")


def mpfr_to_base32(value: mpfr, precision_digits: int = 0) -> str:
    """
    Convert a gmpy2 mpfr object to MPFR base-32 string format in decimal notation.
    Outputs format similar to C implementation: "0.g", "a", "100", etc.
    
    Args:
        value: gmpy2 mpfr value to convert
        precision_digits: Number of base-32 digits to generate (0 = automatic based on precision)
    
    Returns:
        Base-32 string in decimal notation
    """
    if value == 0:
        return "0"
    
    # Use gmpy2's digits() function to get base-32 representation
    # This directly calls MPFR's mpfr_get_str, just like the C code
    # digits() returns (mantissa_str, exponent, precision_used)
    mantissa_str, exp, _ = value.digits(32, precision_digits)
    
    # Handle special cases
    if mantissa_str in ['@NaN@', '@Inf@', '-@Inf@']:
        return mantissa_str
    
    # mantissa_str is the mantissa without decimal point
    # exp is the exponent (number of digits before decimal point)
    sign = '-' if mantissa_str.startswith('-') else ''
    if sign:
        mantissa_str = mantissa_str[1:]  # Remove sign from mantissa
    
    mantissa_len = len(mantissa_str)
    
    if exp > 0:
        # Positive exponent: digits before decimal point
        if exp >= mantissa_len:
            # Need trailing zeros (e.g., "100" when mantissa is "1" and exp is 3)
            result = mantissa_str + '0' * (exp - mantissa_len)
        else:
            # Insert decimal point (e.g., "12.34" when mantissa is "1234" and exp is 2)
            integer_part = mantissa_str[:exp]
            fractional_part = mantissa_str[exp:]
            result = integer_part + '.' + fractional_part
    elif exp == 0:
        # All digits after decimal point, starting with non-zero
        result = '0.' + mantissa_str
    else:
        # Negative exponent: leading zeros after decimal point
        result = '0.' + '0' * (-exp) + mantissa_str
    
    result = sign + result
    # Remove trailing zeros before returning
    return remove_trailing_zeros(result)


def decimal_to_mpfr_base32(d, precision_bits: int = 256) -> str:
    """
    Convert a numeric value to MPFR base-32 string format.
    
    Args:
        d: Numeric value to convert (int, float, str, or mpfr)
        precision_bits: Precision in bits for the conversion (default: 256)
    
    Returns:
        Base-32 string in decimal notation
    """
    # Convert to mpfr with specified precision
    with gmpy2.context(precision=precision_bits):  # type: ignore
        if isinstance(d, mpfr):
            value = d
        else:
            value = mpfr(str(d))
        
        # Calculate number of base-32 digits to generate
        # Formula: digits = ceil(bits * log(2) / log(32)) + 1
        import math
        precision_digits = math.ceil(precision_bits * math.log(2) / math.log(32)) + 1
        
        return mpfr_to_base32(value, precision_digits)

File: c_cal/mpfr_base32.c
#include "mpfr_base32.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * Remove trailing zeros from a number string that has a decimal point
 * Also removes the decimal point if no fractional part remains
 */
static void remove_trailing_zeros(char *str) {
    // Find decimal point
    char *decimal = strchr(str, '.');
    if (decimal == NULL) {
        return;  // No decimal point, nothing to do
    }
    
    // Find end of string
    char *end = str + strlen(str) - 1;
    
    // Remove trailing zeros
    while (end > decimal && *end == '0') {
        *end = '\0';
        end--;
    }
    
    // Remove decimal point if it's now at the end
    if (*end == '.') {
        *end = '\0';
    }
}

/**
 * Parse a base-32 string to MPFR number
 */
int parse_base32_to_mpfr(const char *str, mpfr_t result, mpfr_prec_t prec) {
    mpfr_set_prec(result, prec);
    return mpfr_set_str(result, str, BASE, MPFR_RNDN);
}

/**
 * Convert MPFR number to base-32 string
 */
char* mpfr_to_base32(mpfr_t value) {
    mpfr_exp_t exp;
    char *str = mpfr_get_str(NULL, &exp, BASE, 0, value, MPFR_RNDN);
    
    if (str == NULL) {
        return NULL;
    }
    
    // Handle special cases
    if (strcmp(str, "@NaN@") == 0 || strcmp(str, "@Inf@") == 0 || strcmp(str, "-@Inf@") == 0) {
        char *result = malloc(strlen(str) + 1);
        if (result) {
            strcpy(result, str);
        }
        mpfr_free_str(str);
        return result;
    }
    
    // Handle zero
    if (mpfr_zero_p(value)) {
        mpfr_free_str(str);
        char *result = malloc(2);
        if (result) {
            strcpy(result, "0");
        }
        return result;
    }
    
    // mpfr_get_str returns mantissa and exponent where:
    // - exp is the number of digits before the radix point
    // - If exp > 0: integer part has 'exp' digits
    // - If exp <= 0: number is 0.000...mantissa (with -exp leading zeros after point)
    
    size_t len = strlen(str);
    int sign_offset = (str[0] == '-') ? 1 : 0;
    size_t mantissa_len = len - sign_offset;
    
    // Calculate result size
    size_t result_size;
    if (exp > 0) {
        if ((size_t)exp >= mantissa_len) {
            // Need trailing zeros: mantissa000
            result_size = len + (exp - mantissa_len) + 10;
        } else {
            // Need decimal point: mantissa -> man.tissa
            result_size = len + 10;
        }
    } else {
        // Need leading zeros: 0.000mantissa
        result_size = len + (-exp) + 10;
    }
    
    char *result = malloc(result_size);
    if (result == NULL) {
        mpfr_free_str(str);
        return NULL;
    }
    
    // Build the result string
    char *dest = result;
    char *src = str;
    
    // Copy sign if present
    if (sign_offset) {
        *dest++ = *src++;
    }
    
    if (exp > 0) {
        if ((size_t)exp >= mantissa_len) {
            // Copy all mantissa digits
            strcpy(dest, src);
            dest += mantissa_len;
            // Add trailing zeros
            for (long i = mantissa_len; i < exp; i++) {
                *dest++ = '0';
            }
            *dest = '\0';
        } else {
            // Insert decimal point
            strncpy(dest, src, exp);
            dest += exp;
            *dest++ = '.';
            strcpy(dest, src + exp);
        }
    } else {
        // Leading zeros: 0.000...mantissa
        *dest++ = '0';
        *dest++ = '.';
        for (long i = 0; i < -exp; i++) {
            *dest++ = '0';
        }
        strcpy(dest, src);
    }
    
    // Remove trailing zeros before returning
    remove_trailing_zeros(result);

    mpfr_free_str(str);
    return result;
}